46_revocation_severity_and_context.md

```diff
@@ -0,0 +1,301 @@
+levels of events and adds the option to provide more context for events.
+
+All parts that cause revocation events are tagged with an unique id and against
+those ids user scan specify a severity level and provide some context information.
+
+Instead of stop polling an agent if a failure occurred the agent is added back
+for checking if the failure is recoverable. If it is not a revocation event with
+the highest severity is generated.
+
+All the send events are visible in the verifier API for other systems to use.
+
+Keylime also gains the ability to specify context for a revocation event.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this enhancement is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+The option to classify the revocation events opens the possibility that events
+are not handled. To mitigate this all revocation events that are not explicitly
+from the user classified are assigned the highest severity level.
+
+The same concept applies for revocation events that are caused failures that
+make the agent irrecoverable. Those also always need to be treated with the
+highest severity by default because otherwise some events wont be caught is a
+irrecoverable failure is triggered before.
+
+## Design Details
+We keep the current model of the states, but modify the behavior of the failure
+states. If we are in a failure state that is recoverable the polling of
+the agent is stopped otherwise the agent is still added back for normal polling.
+
+A new failure object will be introduced which holds the all the revocation
+events that are being produced by the checks and if any of the failures makes
+the agent not recoverable. Irrecoverable events currently are retry timeouts and
+quote validation.
+
+All parts of the validation process which such as (`check_qoute` and
+`check_pcrs`) will append all events to that object instead of returning false.
+If the validation process advances without a step that failed the failure object
+must be marked as irrecoverable and only then the function can return without
+validating further. The state of the agent after that should be
+`QUOTE_FAILED_IRRECOVERABLE`.
+
+The highest severity is 0 and the higher the number the number the lower the
+severity. This makes it easy to add less sever levels without the need to
+increase the level of the highest severity.
+
+A new table `revocation_events` is introduced to save the already sent
```

Comment:
The existence of a new table worries me, to be honest. Would this table continuously grow within the lifetime of an agent? In other words, are we expecting multiple lines with the same agent id?

Reply:
Yes we would record every event that was generated for a agent during its lifetime. 
This allows us to do two things. First can provide an API entry point for all the revocation events that have been generated for other systems. Second we want to check if we already sent an event. For example if a static PCR is failing it will in most cases always fail again and we want to be only notified once. Always resending the event would cause potentially a lot of redundant traffic.

We could put that information into a JSON object and store it string encoded in one single entry (if the database allows arbitrary length for fields), but that would make checking if we send a event already always a linear action. 
Another approach is to only store the latest or the one with the current highest severity, but this will remove the option to check if a event was sent and a event log for an agent must be recorded by a component outside of Keylime.

Reply:
This proposal would be breaking a key assumption on Keylime databases so far: these are not "historical", i.e., they do not accumulate data about an `agent` and, if destroyed, it can be quickly rebuilt from the "ground truth", which is the current state requested from such `agents`. This is design point made before my time as one of the keylime maintainers, but I must say that it is a point worth preserving. This being said, I would vote against this enhancement based due to the need for this change.

Reply:
If changing that is not an option, what is your opinion on just saving the highest severity level and sending a event only if it has a higher severity? This would not require an extra table and cloud be more easily restored by requesting the state from the agent. Now the clients receiving the events can only make decisions on based on the severity level but I think in most cases this will be still a valuable improvement.

I think that it will be very useful to have some kind of API (either over REST API or Python module) such that external systems can get more information about all revocations events out of Keylime. Those systems can than implement the record keeping outside of Keylime and relay the necessary information to the relevant systems.
(We are already implementing a bridge between Keylime and our system for the tenant functionality, so we could move that out of Keylime and into our bridge.)

Reply:
My short answer would be yes, but I believe your original proposal, the "provenance" of the component/stage where the failure happened is tremendously useful on itself. After several months using Keylime in Production, I became a firm believer in the need to "decouple" the revocation from the core of keylime (the current implementation is not particularly good, flexible or robust) and adopting a more configurable approach (e.g., execute a given command in case an agent fails, passing all the appropriate parameters) is the way to move forward.

Reply:
I would propose that I keep the tracking where a event was generated and the user specified rules in this proposal but restructure the event sending part to just use the labels to demonstrate some aspects of the new gained capabilities of Keylime.

For the redesign of the revocation mechanism of Keylime we then open a new enhancement proposal which uses the added capabilities from this proposal as a base. (Here I would propose some kind of plugin API where the current revocation mechanism is just a plugin and can be swapped out for other mechanisms.)

Reply:
I agree with @maugustosilva here in that we shouldn't have keylime keep track of events. Having an extra status field about the last failure, etc attached to the verifier table would be good. Having an extra table that just grows over time would not be. We do need to tackle the idea of an event log and how to store/query/search/maintain that long term, but that's better suited as APIs to existing external systems like adapters for elasticsearch or similar.